### ISSUE TITLE
Border StrokeShape being forgotten (?)

### ISSUE THREAD

**Issue by Mysame** [AdamEssenmacher/MemoryToolkit.Maui#18]
Behold a sample of our logging, I'm sure you get the gist.

![image](https://github.com/AdamEssenmacher/MemoryToolkit.Maui/assets/1966231/9ca6e482-64cd-4551-8558-2903cf02fd99)

We do use RoundRectangle to specify the border shape in the xaml where we've implemented the TearDown 
```
<CollectionView.ItemTemplate>
  <DataTemplate>
     <Border Margin="10,5,10,5"
             Padding="0"
             BackgroundColor="White">

         <Border.StrokeShape>
             <RoundRectangle CornerRadius="15" />
         </Border.StrokeShape>
```

So I suppose there's a very real possibility of the parent being disposed whilst the RoundRectangle is still alive.

**Comment by sjorsmiltenburg** [comment]
Good find, did you also report this on the dotnet/maui github?

**Comment by Mysame** [comment]
I was considering it because the urgency of the output message, but I figured there's no real point poking them if it's most likely caused by this toolkit. 
But if requested I definitely will, however their response will most likely be "well stop garbage collecting `RealParent`, then".

**Comment by sjorsmiltenburg** [comment]
Fair point. But I would say let them analyse the cause and determine if the relation with the toolkit is at fault.

To be clear I don't have in depth knowledge of the inner workings of this memorytoolkit.
I am just a developer that applauds all efforts in finding / fixing memoryleaks in Xamarin / Maui.
I want all the Maui.Windows (and other platform) memory leaks fixed ASAP as it currently does not seem possible to develop a stable product on MAUI (or xamarin for that matter). As I have a product that needs to be able to run 24/7 this is a big problem for me.

**Comment by AdamEssenmacher** [comment]
The error `The RealParent on...` was introduced in https://github.com/dotnet/maui/pull/22561, which shipped with SR6. This PR should fix the propagating leak behavior this toolkit was primarily designed to address!

So yes, using this toolkit is probably causing this (new) warn log to occur--but the good news is that TearDownBehavior probably isn't needed anymore. Try removing it and see if any leaks still happen with `LeakMonitorBehavior`.

**Comment by Mysame** [comment]
That mentions iOS specifically, I should have mentioned this is on Android.

Disabling the `TearDownBehavior` allows memory to rise steadily in our app before it crashes, something that doesn't happen (as fast) with it enabled.

**Comment by AdamEssenmacher** [comment]
> That mentions iOS specifically, I should have mentioned this is on Android.

The PR comments do not fully explain the scope of the impact here. I'm not sure if PureWeen realized it or not. The reason leaks were spreading through MAUI pages so rampantly was because the strong references up through the visual tree via the `Parent` property. By making this property a weak reference, leaks should be compartmentalizing at the lowest level they occur (...I think; I haven't had a chance to validate this in practice yet).

> Disabling the `TearDownBehavior` allows memory to rise steadily in our app before it crashes, something that doesn't happen (as fast) with it enabled.

That's good to know. `TearDownBehavior` does a handful of things, and it sounds like some of them are still useful. In addition to breaking apart views (which I think is the part that's no longer necessary), it also clears binding contexts and calls `DisconnectHandler()`, which isn't _supposed_ to be necessary, but sometimes is.

Anyway, until I have a chance to dive back into this (I had to retreat back into Xamarin for a bit) you might try using the `Suppress` option to exclude the Border from the effects of `TearDownBehavior` to maybe avoid those noisy logs.

**Comment by any2info** [comment]
First of all @AdamEssenmacher thanks for this!!! We have been struggling with migration to MAUI for a few months now and are finally at the point where we can start addressing performance related issues. 

Using the nuget i'm getting the same issue as described here. In my case a Border using in a control template defined globally in app.xml. While examining this i encountered something weird, which is i set the border shape in xaml the message no longer is generated. If i however let it be definied by default from my style (also definied in app.xaml) it does show, with the exact same value StrokeShape="RoundRectangle 5".

Also i can't really determine from this thread if this is an actual leak or issue in the toolkit?

**Comment by AdamEssenmacher** [comment]
My [comment above](https://github.com/AdamEssenmacher/MemoryToolkit.Maui/issues/18#issuecomment-2206806348) is probably still relevant. This toolkit was developed against MAUI 8 SR3/4. `TearDownBehavior` achieves its goals by actively working against MAUI's design intent, allowing (or even encouraging) objects to be eligible for garbage collection before MAUIs design would expect. SR 6 introduced this logging message when objects get collected too aggressively.

My advice:

- Confirm whether or not you actually have a leak. If there isn't a leak, there is no reason to use `TearDownBehavior` and the annoying logs go away.
- If you do have a leak, apply TearDownBehavior at the lowest level needed to eliminate it. I get it that can seem easier to to slap it on the top of every page, but this can have unintended consequences (like these annoying logs). Used most ideally, `TearDownBehavior.Cascade` and `TearDownBehavior.Suppress` should be used together like a scalpel to target specific, known leaks as-needed. 
- If TearDownBehavior is actually needed to eliminate a leak, and this also causes the annoying logs, confirm whether or not there are any other negative effects. You might just have to choose between having a leak vs having annoying logs.

@any2info https://github.com/dotnet/maui/issues/20163 you might find that issue interesting.

**Comment by cagriy** [comment]
@AdamEssenmacher , thanks for this great tool.

Somewhat similar to the issue mentioned here, if I set RoundRectangle StrokeShape for Border in global styles, the toolkit reports a leak, although when I perform a gcdump everything looks clean. (.NET 9.0.21)

I am guessing that this is a good candidate for using suppress. If so, is there a practical way to suppress only the leaks referring to RoundRectangle (or Borders) ? The page has 10+ Borders. Or should I name and suppress each border individually?

**Comment by AdamEssenmacher** [comment]
>  if I set RoundRectangle StrokeShape for Border in global styles, the toolkit reports a leak, although when I perform a gcdump everything looks clean. 

I'd be interested in seeing a small sample project on this.  The toolkit only reports leaks for objects that are still in memory when they're not expected to be, which means one of three things is true:

1) The lifecycle of your views is outside of the definition of 'done with' outlined in the project readme.
or
2) The toolkit is not getting the GC to behave deterministically.
or
3) There's a misunderstanding around gcdump

> is there a practical way to suppress only the leaks referring to RoundRectangle (or Borders) ?

I don't think so.

**Comment by cagriy** [comment]
Hi @AdamEssenmacher. Please find the sample project here: https://github.com/cagriy/RoundRectangle

Regarding gcdump, with these kinds of leaks, when I get a dump and search for either RoundRectangle or SecondPage, I typically find them left behind. On this occasion, SecondPage and RoundRectangle seem to be cleaned up properly.

**Comment by AdamEssenmacher** [comment]
Thanks for the sample project @cagriy.

I think the answer is (1) from my last comment. When you set the `StrokeShape` property in a global style to an instance of `RoundRectangle` like this, the global `RoundRectagle` instance ends up in your page's visual tree, and so it does not get GC'ed when the page goes out of scope because the style still has a reference to it.

The `Suppress` property exists for this reason--to exclude a view we don't expect to be collected when its parent page goes out of scope. For example:

```
    <Style TargetType="Border">
        <Setter Property="StrokeShape">
            <Setter.Value>
                <RoundRectangle CornerRadius="10"
                                mtk:LeakMonitorBehavior.Suppress="True" />
            </Setter.Value>
        </Setter>
    </Style>
```
